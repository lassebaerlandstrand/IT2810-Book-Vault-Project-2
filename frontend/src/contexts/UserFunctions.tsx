import { useContext } from 'react';
import { UserContext } from './UserContext';

/**
 * Custom hook to access the UserContext.
 * @returns {UserContextProps} The user context value containing user info and setter
 *     - info: The user info object
 *     - setUser: Function to set the user info
 * @throws {Error} If used outside of UserProvider
 */
export const useUser = () => {
  const context = useContext(UserContext);
  if (!context) {
    throw new Error('useUser must be used within a FavoritesProvider');
  }
  return context;
};
