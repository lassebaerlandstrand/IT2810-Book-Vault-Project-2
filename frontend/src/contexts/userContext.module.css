.centeredOnPage {
  height: 100vh;
}
