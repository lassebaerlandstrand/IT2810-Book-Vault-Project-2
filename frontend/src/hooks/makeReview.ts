import { useMutation } from '@apollo/client';
import { CreateReviewInput } from '@/generated/graphql';
import { CREATE_REVIEW } from '@/graphql/mutations/reviews';

/**
 * Custom hook that handles review creation functionality.
 * @returns Object containing:
 * - submitReview: Function to create a new review
 * - loading: Boolean indicating if the mutation is in progress
 * - error: Any error that occurred during the mutation
 */
export const makeReview = () => {
  // Return the mutation and its states (data, loading, error)
  const [createReview, { data, loading, error }] = useMutation(CREATE_REVIEW);

  const submitReview = async (input: CreateReviewInput) => {
    try {
      await createReview({
        variables: { input },
        update: (cache, { data: mutationData }) => {
          if (mutationData?.createReview.success) {
            // On update remove the bookReviews "focusUserUUID" queries from cache as
            // they are now stale
            const allKeys = cache.extract().ROOT_QUERY;

            Object.keys(allKeys).forEach((key) => {
              if (key.startsWith('bookReviews({"focusUserUUID"')) {
                const jsonString = key.replace('bookReviews(', '').replace(')', '');
                const parsed = JSON.parse(jsonString);
                const { limit, offset } = parsed;

                cache.evict({
                  id: 'ROOT_QUERY',
                  fieldName: 'bookReviews',
                  args: {
                    focusUserUUID: input.userUUID,
                    limit,
                    offset,
                  },
                });
              }
            });

            cache.gc();
          }
        },
      });
    } catch (e) {
      console.error('Error during update:', e);
    }
  };

  return {
    submitReview, // Function to trigger the mutation later
    success: data?.createReview.success ?? !error,
    message: data?.createReview.message ?? (error && 'Error making review. Try again later.'),
    loading,
    error,
  };
};
