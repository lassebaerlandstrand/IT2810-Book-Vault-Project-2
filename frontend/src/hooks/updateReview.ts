//UPDATE_REVIEW

import { useMutation } from '@apollo/client';
import { UpdateReviewInput } from '@/generated/graphql';
import { UPDATE_REVIEW } from '@/graphql/mutations/reviews';

export const updateReview = () => {
  const [update, { data, loading, error }] = useMutation(UPDATE_REVIEW, {
    update(cache) {
      // On update remove the bookReviews "focusUserUUID" queries from cache as
      // they are now stale
      const allKeys = cache.extract().ROOT_QUERY;

      Object.keys(allKeys).forEach((key) => {
        if (key.startsWith('bookReviews({"focusUserUUID"')) {
          cache.evict({
            id: cache.identify({
              __typename: 'Query',
              key,
            }),
          });
        }
      });

      //Clean up
      cache.gc();
    },
  });

  const submitUpdate = async (input: UpdateReviewInput) => {
    try {
      await update({
        variables: { input },
      });
    } catch (e) {
      console.error('Error during update:', e);
    }
  };

  return {
    submitUpdate,
    success: data?.updateReview.success,
    message: data?.updateReview.message,
    loading,
    error,
  };
};
