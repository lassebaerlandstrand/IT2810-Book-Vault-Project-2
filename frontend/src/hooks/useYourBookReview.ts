import { useQuery } from '@apollo/client';
import { GET_YOUR_BOOK_REVIEW } from '@/graphql/queries/reviews';

type UseBooksReviewsArgs = {
  bookID: string;
  userUUID: string;
};

/**
 * Custom hook that fetches a user's review for a specific book
 * @param {string} bookID - The ID of the book to fetch the review for
 * @param {string} userUUID - The UUID of the user to fetch the review for
 * @returns Object containing:
 * - review: The review data if found
 * - loading: Boolean indicating if the query is in progress
 * - error: Any error that occurred during the query
 * - refetch: Function to refetch the review
 */
export const useYourBookReview = ({ bookID, userUUID }: UseBooksReviewsArgs) => {
  const { data, loading, error, refetch } = useQuery(GET_YOUR_BOOK_REVIEW, {
    variables: { bookID, userUUID },
  });

  return {
    review: data?.bookReview || undefined,
    loading,
    error,
    refetch,
  };
};
