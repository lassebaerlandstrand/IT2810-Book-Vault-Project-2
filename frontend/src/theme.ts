import {
  Container,
  createTheme,
  DEFAULT_THEME,
  Loader,
  MantineColorsTuple,
  rem,
} from '@mantine/core';
import { BookLoader } from './components/Loading/Loading';

const CONTAINER_SIZES: Record<string, string> = {
  xxs: '24em',
  xs: '30em',
  sm: '48em',
  md: '64em',
  lg: '75em',
  xl: '90em',
};

// https://colorkit.co/color-palette-generator/02070b-030a11-030d16-051321-06192b-092641-0c3256-0f3f6c-124b81-155896/
const darkBlue: MantineColorsTuple = [
  '#155896',
  '#124b81',
  '#0f3f6c',
  '#0c3256',
  '#092641',
  '#06192b',
  '#051321',
  '#030d16',
  '#030a11',
  '#02070b',
];

// https://colorkit.co/color-palette-generator/6a5f4a-d4bd94-dac5a1-e2d1b5-e9ddc8-eee4d3-f2eade-f6f0e7-f9f5ef-fcfaf7/
const beige: MantineColorsTuple = [
  '#fcfaf7',
  '#f9f5ef',
  '#f6f0e7',
  '#f2eade',
  '#eee4d3',
  '#e9ddc8',
  '#e2d1b5',
  '#dac5a1',
  '#d4bd94',
  '#6a5f4a',
];

export const theme = createTheme({
  fontFamily: `Poppins, ${DEFAULT_THEME.fontFamily}`,
  breakpoints: {
    xxs: '24em', // 384px
    xs: '30em', // 480px
    sm: '48em', // 768px
    md: '64em', // 1024px
    lg: '75em', // 1200px
    xl: '90em', // 1440px
  },
  colors: {
    'dark-blue': darkBlue,
    beige,
  },
  components: {
    Container: Container.extend({
      vars: (_, { size, fluid }) => ({
        root: {
          '--container-size': fluid
            ? '100%'
            : size !== undefined && size in CONTAINER_SIZES
              ? CONTAINER_SIZES[size]
              : rem(size),
        },
      }),
    }),
    Loader: Loader.extend({
      defaultProps: {
        loaders: { ...Loader.defaultLoaders, book: BookLoader },
      },
      vars: (_, props) => {
        if (props.size === 'xxl') {
          return {
            root: {
              '--loader-size': rem(90),
            },
          };
        }
        return { root: {} };
      },
    }),
  },
});
