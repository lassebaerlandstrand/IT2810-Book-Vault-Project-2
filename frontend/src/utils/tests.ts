/**
 * Removes random attributes generated by Mantine from DOM elements.
 * This is necessary to ensure consistent snapshots in tests, as Mantine uses random IDs.
 */
export const removeMantineRandomAttributes = () => {
  const attributesToRemove = [
    ...document.body.querySelectorAll('div [id^="mantine"]'),
    ...document.body.querySelectorAll('div [for^="mantine"]'),
  ];
  attributesToRemove.forEach((element) => {
    element.removeAttribute('for');
    element.removeAttribute('id');
    element.removeAttribute('class');
    element.removeAttribute('aria-describedby');
  });
};
