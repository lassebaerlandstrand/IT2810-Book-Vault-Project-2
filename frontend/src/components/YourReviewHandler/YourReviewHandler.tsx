import { useEffect, useState } from 'react';
import { ApolloQueryResult, OperationVariables } from '@apollo/client';
import { Button, Divider, Flex, Grid, Rating, Text, Textarea } from '@mantine/core';
import { useUser } from '@/contexts/UserFunctions';
import { Book } from '@/generated/graphql';
import { makeReview } from '@/hooks/makeReview';
import { updateReview } from '@/hooks/updateReview';
import { useYourBookReview } from '@/hooks/useYourBookReview';
import ReviewStack from '../ReviewStack/ReviewStack';

type ReviewProps = {
  book: Book;
  updateAvgRating: (variables?: Partial<OperationVariables>) => Promise<ApolloQueryResult<any>>;
};

const YourReviewHandler = ({ book, updateAvgRating }: ReviewProps) => {
  const [visible, setVisible] = useState(false);
  const [rating, setRating] = useState(1);
  const [text, setText] = useState('');

  const UUID: string = useUser().info.UUID;

  // Fetch your review of the book (if it exists)
  const { review: yourReview, refetch: refetchYourReview } = useYourBookReview({
    bookID: book.id,
    userUUID: UUID,
  });

  // For submitting a review (NR = new review)
  const { submitReview, updatedRating: updatedRatingNR, loading: yourReviewLoading } = makeReview();

  // For updating reviews (UR = updated review)
  const {
    submitUpdate,
    updatedRating: updatedRatingUR,
    loading: loadingUpdateReview,
  } = updateReview();

  // Toggle the review
  const toggleReviewDisplay = () => {
    setVisible((prev) => !prev);
  };

  // Cancel update to review
  const cancel = () => {
    if (yourReview) {
      setRating(yourReview.rating);
      setText(yourReview.description);
    }
    setVisible(false);
  };

  // Update review
  const update = () => {
    if (yourReview) {
      setVisible(false);

      submitUpdate({
        reviewUUID: yourReview.UUID,
        description: text,
        rating,
      });

      refetchYourReview();
    }
  };

  // Submit review
  const submit = () => {
    setVisible(false);
    submitReview({
      userUUID: UUID,
      bookID: book.id,
      description: text,
      rating,
    });
    refetchYourReview();
  };

  // Update rating + refetch your rating
  const updateRating = (updatedRating: number) => {
    if (!updatedRating) {
      return;
    }
    if (updatedRating !== -1) {
      updateAvgRating();
    }
  };

  // Refetch your review after either posting one or updating it
  useEffect(() => {
    updateRating(updatedRatingNR);
  }, [updatedRatingNR]);

  useEffect(() => {
    updateRating(updatedRatingUR);
  }, [updatedRatingUR]);

  // For updating reviews
  useEffect(() => {
    if (yourReview) {
      setRating(yourReview.rating);
      setText(yourReview.description);
    }
  }, [yourReview]);

  return (
    <>
      {!visible && !yourReviewLoading && !loadingUpdateReview ? (
        <Grid>
          <Grid.Col span="auto">
            <Button variant="filled" onClick={toggleReviewDisplay}>
              {!yourReview ? 'Give Review' : 'Edit Review'}
            </Button>
          </Grid.Col>
        </Grid>
      ) : (
        <></>
      )}

      {visible ? (
        <>
          <Text size="sm">Your rating</Text>
          <Rating size="xl" value={rating} onChange={setRating} />

          <Text size="sm">Your review</Text>

          <Textarea value={text} onChange={(event) => setText(event.currentTarget.value)} />

          <Flex justify="Right" gap="lg">
            <Button variant="filled" color="gray" onClick={cancel}>
              Cancel
            </Button>
            {yourReview ? (
              <Button variant="filled" onClick={update}>
                Update Review
              </Button>
            ) : (
              <Button variant="filled" onClick={submit}>
                Submit Review
              </Button>
            )}
          </Flex>
        </>
      ) : (
        <></>
      )}

      {!visible && yourReview ? (
        <>
          <ReviewStack
            reviews={[
              {
                UUID: yourReview.UUID,
                at: yourReview.at,
                rating: yourReview.rating,
                description: yourReview.description,
                user: {
                  name: 'Your review',
                  UUID,
                },
              },
            ]}
            type="yourReview"
          />
          <Divider size="xs" label="Other reviews" labelPosition="center" />
        </>
      ) : (
        <></>
      )}
    </>
  );
};

export default YourReviewHandler;
