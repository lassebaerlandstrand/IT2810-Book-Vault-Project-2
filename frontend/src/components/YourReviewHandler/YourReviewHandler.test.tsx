import { MockedProvider } from '@apollo/client/testing';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { MantineProvider } from '@mantine/core';
import { useUser } from '@/contexts/UserFunctions';
import { makeReview } from '@/hooks/makeReview';
import { updateReview } from '@/hooks/updateReview';
import { useYourBookReview } from '@/hooks/useYourBookReview';
import { theme } from '@/theme';
import { removeMantineRandomAttributes } from '@/utils/tests';
import { dummyBook } from '../../../test-utils/testVars';
import YourReviewHandler from './YourReviewHandler';

vi.mock('@/contexts/UserFunctions', () => ({
  useUser: vi.fn(),
}));

vi.mock('@/hooks/makeReview', () => ({
  makeReview: vi.fn(),
}));

vi.mock('@/hooks/updateReview', () => ({
  updateReview: vi.fn(),
}));

vi.mock('@/hooks/useYourBookReview', () => ({
  useYourBookReview: vi.fn(),
}));

const dummyUser = {
  info: {
    name: 'You',
    UUID: 'UUID1',
  },
};

const dummyReview = {
  UUID: 'review-1',
  rating: 2,
  description: 'Great book!',
  at: '2023-10-30T12:00:00Z',
};

const customRender = () => {
  return render(
    <MockedProvider addTypename={false}>
      <MantineProvider theme={theme}>
        <MemoryRouter>
          <YourReviewHandler book={dummyBook} />
        </MemoryRouter>
      </MantineProvider>
    </MockedProvider>
  );
};

describe('YourReviewHandler Component', () => {
  beforeEach(() => {
    (useUser as jest.Mock).mockReturnValue(dummyUser);
    (makeReview as jest.Mock).mockReturnValue({
      submitReview: vi.fn(),
      updatedRating: undefined,
      loading: false,
    });
    (updateReview as jest.Mock).mockReturnValue({
      submitUpdate: vi.fn(),
      updatedRating: undefined,
      loading: false,
      error: null,
    });
    (useYourBookReview as jest.Mock).mockReturnValue({
      review: undefined,
      refetch: vi.fn(),
      loading: false,
      error: null,
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('that "Give Review" and "Cancel" buttons work', async () => {
    customRender();

    // Check that clicking the Give Review and cancel works properly
    expect(screen.getByText('Give Review')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Give Review'));

    expect(screen.getByText('Submit Review')).toBeInTheDocument();
    expect(screen.getByText('Cancel')).toBeInTheDocument();
    expect(screen.getByText('Your rating')).toBeInTheDocument();
    expect(screen.getByText('Your review')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Cancel'));
    expect(screen.getByText('Give Review')).toBeInTheDocument();
  });

  test('"Edit Review" option shows up when you have made a review', async () => {
    (useYourBookReview as jest.Mock).mockReturnValue({
      review: dummyReview,
      refetch: vi.fn(),
      loading: false,
      error: undefined,
    });

    customRender();

    expect(screen.getByText('Edit Review')).toBeInTheDocument();
  });

  test('"Cancel" reverts changes you made', async () => {
    (useYourBookReview as jest.Mock).mockReturnValue({
      review: dummyReview,
      refetch: vi.fn(),
      loading: false,
      error: undefined,
    });

    customRender();

    expect(screen.getByText('Edit Review')).toBeInTheDocument();
    fireEvent.click(screen.getByText('Edit Review'));

    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'Updated review text' } });

    expect(screen.getByText('Cancel')).toBeInTheDocument();
    fireEvent.click(screen.getByText('Cancel'));
    expect(screen.getByText(dummyReview.description)).toBeInTheDocument();
  });

  test('submits a new review', async () => {
    const submitReviewMock = vi.fn();
    (makeReview as jest.Mock).mockReturnValue({ submitReview: submitReviewMock, loading: false });

    customRender();

    fireEvent.click(screen.getByText('Give Review'));
    fireEvent.change(screen.getByRole('textbox'), { target: { value: 'A great one!' } });
    fireEvent.click(screen.getByText('Submit Review'));

    await waitFor(() => {
      expect(submitReviewMock).toHaveBeenCalledWith({
        userUUID: dummyUser.info.UUID,
        bookID: dummyBook.id,
        description: 'A great one!',
        rating: 1,
      });
    });
  });

  it('matches snapshot', () => {
    const { asFragment } = customRender();

    removeMantineRandomAttributes();
    expect(asFragment()).toMatchSnapshot();
  });
});
