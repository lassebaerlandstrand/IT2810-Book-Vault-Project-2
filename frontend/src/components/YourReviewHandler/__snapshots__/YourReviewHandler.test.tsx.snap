// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`YourReviewHandler Component > matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-font-family-headings: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-breakpoint-xs: 30em;--mantine-breakpoint-md: 64em;--mantine-breakpoint-xl: 90em;--mantine-breakpoint-xxs: 24em;--mantine-color-logo-theme-0: #fefefe;--mantine-color-logo-theme-1: #cbcfd2;--mantine-color-logo-theme-2: #97a0a6;--mantine-color-logo-theme-3: #7d8991;--mantine-color-logo-theme-4: #63717b;--mantine-color-logo-theme-5: #495a65;--mantine-color-logo-theme-6: #2f434f;--mantine-color-logo-theme-7: #132937;--mantine-color-logo-theme-8: #0b202e;--mantine-color-logo-theme-9: #061017;}:root[data-mantine-color-scheme="dark"]{--mantine-color-logo-theme-text: var(--mantine-color-logo-theme-4);--mantine-color-logo-theme-filled: var(--mantine-color-logo-theme-8);--mantine-color-logo-theme-filled-hover: var(--mantine-color-logo-theme-9);--mantine-color-logo-theme-light: rgba(47, 67, 79, 0.15);--mantine-color-logo-theme-light-hover: rgba(47, 67, 79, 0.2);--mantine-color-logo-theme-light-color: var(--mantine-color-logo-theme-3);--mantine-color-logo-theme-outline: var(--mantine-color-logo-theme-4);--mantine-color-logo-theme-outline-hover: rgba(99, 113, 123, 0.05);}:root[data-mantine-color-scheme="light"]{--mantine-color-logo-theme-text: var(--mantine-color-logo-theme-filled);--mantine-color-logo-theme-filled: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-filled-hover: var(--mantine-color-logo-theme-7);--mantine-color-logo-theme-light: rgba(47, 67, 79, 0.1);--mantine-color-logo-theme-light-hover: rgba(47, 67, 79, 0.12);--mantine-color-logo-theme-light-color: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-outline: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-outline-hover: rgba(47, 67, 79, 0.05);}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 29.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 30em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 63.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 64em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 89.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 90em) {.mantine-hidden-from-xl {display: none !important;}}@media (max-width: 23.99375em) {.mantine-visible-from-xxs {display: none !important;}}@media (min-width: 24em) {.mantine-hidden-from-xxs {display: none !important;}}
  </style>
  <style
    data-mantine-styles="inline"
  >
    .__m__-r8k{--grid-gutter:var(--mantine-spacing-md);}
  </style>
  <div
    class="m_410352e9 mantine-Grid-root __m__-r8k"
  >
    <div
      class="m_dee7bd2f mantine-Grid-inner"
    >
      <style
        data-mantine-styles="inline"
      >
        .__m__-r8m{--col-flex-grow:1;--col-flex-basis:0rem;--col-max-width:100%;}
      </style>
      <div
        class="m_96bdd299 mantine-Grid-col __m__-r8m"
      >
        <button
          class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
          data-variant="filled"
          style="--button-bg: var(--mantine-color-blue-filled); --button-hover: var(--mantine-color-blue-filled-hover); --button-color: var(--mantine-color-white); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
          type="button"
        >
          <span
            class="m_80f1301b mantine-Button-inner"
          >
            <span
              class="m_811560b9 mantine-Button-label"
            >
              Give Review
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
